### ISSUE TITLE
Add rollback to Projectiles

### ISSUE THREAD

**Issue by TheYellowArchitect** [foxssake/netfox#312]
### :sparkles: Description

[Projectiles don't work with Rollback](https://foxssake.github.io/netfox/netfox.extras/guides/network-weapon/):

>"Upon firing, sending a request to the server and waiting for the response with the projectile would introduce a delay. Doing a full-on state synchronization with MultiplayerSynchronizer or RollbackSynchronizer can be unfeasible with too many projectiles, and unnecessary, since most of the time, projectiles act and move the same way regardless of their surroundings."

Rollback currently works per entity, using `RollbackSynchronizer`. I agree in the above that each Projectile should **not** have a `RollbackSynchronizer` but what should exist to make them viable for rollback is critical events to track which tick it was created, which tick it was destroyed (this logic can be used for more entities btw)

So, at the `NetworkRollback` level, there would be a `ProjectileHistorian` node, which stores a Dictionary of Projectiles.
Key = Projectile ID (currently 12 character strings), Value = Projectile node itself.
There are also 2 new dictionaries which have as value the projectile ID. These 2 dictionaries hold the creation tick and the destruction/free tick.

### Code Use-Case

A projectile is spawned at tick 57, and explodes at tick 78.
The game is at tick 79, and a rewind happens to tick 75. I assume this would happen if a new input is received for tick 75.

On rewinding to any tick, `ProjectileHistorian` checks if a projectile which no longer exists on this tick (79) exists on the target tick (75) and if so, it instantiates its `PackedScene`.

### Data

Taking the above code use-case in mind, here is what a `ProjectileHistorian` must store for every projectile:
1. Its ID
2. Its `PackedScene` (for instantiating it)
3. Its position, rotation, velocity **for every tick**.
4. Creation tick
5. Death tick

```gdscript
class_name ProjectileData

var projectile_instance: Node
var creation_tick: int
var death_tick: int
var scene_template: PackedScene
```

```gdscript
class_name ProjectileStatus?

var position: Vector3
var rotation: Quaternion
var velocity: Vector3
```

So `ProjectileHistorian` simply has a `Dictionary<ProjectileID, ProjectileData>` and a `Dictionary<ProjectileID, Dictionary<tick, ProjectileStatus>`

Tbf, we could even extrapolate the creation ticks and death ticks completely, and hence probably skip/simplify `ProjectileData`.
The first tick in `Dictionary<ProjectileID, Dictionary<tick, ProjectileStatus>` is the creation tick.
The death tick becomes a function, e.g.
```gdscript

#pseudocode
func get_death_tick() -> int:
    #basically check if it is not freed
    if (projectile.is_in_the_scene == false):
        return the_last_tick_of_the_dictionary
    else:
         return -1
```

In this way, `ProjectileData` is needless. You need 2 primitive dictionaries instead in its place:
```gdscript
var projectiles_stored: Dictionary #<id, Projectile>
var projectiles_scenes: Dictionary #<id, PackedScene>
```

And to add the ticks:
```gdscript
var projectiles_status: Dictionary #<id, Dictionary<tick, ProjectileStatus>>
```

Every name above is liable to change.

It is certain that this `ProjectileHistorian` must hook into some signals from `NetworkRollback`.

### Use-Case

I cannot think of any online game, 2D or 3D not using projectiles of some sort. Metroidvanias, Strategy, FPS, Action Games etc
And given `NetworkWeapon` is included in `netfox.extras`, this ensures projectiles work with the rest of netfox's rollback.

### Distribution

Extension of `NetworkWeapon` logic, so `netfox.extras`

### Notes

Feel free to add any other notes you may have.

**Comment by elementbound** [comment]
> On rewinding to any tick, ProjectileHistorian checks if a projectile which no longer exists on this tick (79) exists on the target tick (75) and if so, it instantiates its PackedScene.

I'm not convinced that destroying and re-instantiating nodes is the best way to go about managing liveness with rollback. Also pulling all projectiles into a centralized object like this goes against the current design philosophy of letting things to individually manage themselves by reacting to central signals. Either way, I'm also not convinced that projectiles should be pulled into a centralized ProjectileHistorian. 

---

On data, this is also very game specific. I'm also not convinced that each projectile needs to have position stored, e.g. if the projectile's movement logic is very simple. Think of RTS games where projectiles get spawned at some location, and then continuously march toward a target unit. You don't need to store location for every tick, you can just calculate the position based on lifetime, starting position and the target unit's position. 

There's also a reason why e.g. RollbackSynchronizer has configurable state properties - when developing netfox, I don't know the game you will be building. And I try to make the least amount of assumptions. Deciding that every projectile must have a position, rotation and velocity breaks that in a big way. 

---

> Extension of `NetworkWeapon` logic, so `netfox.extras`

`NetworkWeapon` does not care about its projectiles much, so I don't see how this would be an extension of it. But projectile logic in general fits in extras, that's true. 

---

So to clarify, you're requesting an example on how to do projectiles with rollback, right?

**Comment by TheYellowArchitect** [comment]
>I'm not convinced that destroying and re-instantiating nodes is the best way to go about managing liveness with rollback.

Pooling can be utilized, which stores like the last X projectiles of each PackedScene :+1: 

>Also pulling all projectiles into a centralized object like this goes against the current design philosophy of letting things to individually manage themselves by reacting to central signals. 

I agree, but how would you implement this feature, when Projectiles themselves are freed?
Even if I hide+disable the projectile on impact, and have it freed only when rollback history limit is reached, it must hook to some central signals somewhere. `NetworkRollback` itself shouldn't be related to Projectiles, so ProjectileHistorian can be used. Though I guess, the logic could move from the ProjectileHistorian to the Projectiles themselves, via some static function.

> I'm also not convinced that each projectile needs to have position stored, e.g. if the projectile's movement logic is very simple. Think of RTS games where projectiles get spawned at some location, and then continuously march toward a target unit. You don't need to store location for every tick, you can just calculate the position based on lifetime, starting position and the target unit's position.

Didn't think of that, sounds like it saves a lot of memory :+1: 

>There's also a reason why e.g. RollbackSynchronizer has configurable state properties - when developing netfox, I don't know the game you will be building. And I try to make the least amount of assumptions. Deciding that every projectile must have a position, rotation and velocity breaks that in a big way.

:+1: 

>So to clarify, you're requesting an example on how to do projectiles with rollback, right?

yup

**Comment by TheYellowArchitect** [comment]
>I'm also not convinced that each projectile needs to have position stored, e.g. if the projectile's movement logic is very simple. Think of RTS games where projectiles get spawned at some location, and then continuously march toward a target unit. You don't need to store location for every tick, you can just calculate the position based on lifetime, starting position and the target unit's position.

>There's also a reason why e.g. RollbackSynchronizer has configurable state properties - when developing netfox, I don't know the game you will be building. And I try to make the least amount of assumptions. Deciding that every projectile must have a position, rotation and velocity breaks that in a big way.

I agree. In my case, I want a grenade, so I do need its position since it bounces off walls/players. So having the position as a state property (not hardcoded) ftw

**Comment by DustieDog** [comment]
> Also pulling all projectiles into a centralized object like this goes against the current design philosophy of letting things to individually manage themselves by reacting to central signals.

Each projectile can be registered to a signal that the proposed ProjectileHistorian emits, which is itself hooked into the NetworkRollback akin to RollbackSynchronizer. That way each projectile is still self-maintained, but has a way to utilize the rollback features without needing a full RollbackSynchronizer node each.

> On data, this is also very game specific. I'm also not convinced that each projectile needs to have position stored

Letting ProjectileHistorian store serialized state data from the projectiles, either through PropertyEntry or something more lightweight, would allow customization of what exactly is stored on a game by game basis.
